xbmc/windowing/android/WinSystemAndroid.cpp

```diff
@@ -224,10 +224,15 @@ void CWinSystemAndroid::InitiateModeChange()
                                     "videoscreen.delayrefreshchange") *
                                 100);
 
-  if (delay < 2000ms)
+  if (delay > 0 && delay < 2000ms)
     delay = 2000ms;
```

Comment:
Might also be applicable with a GUI change, so that user knows the minimum is either 0 or 2 seconds?
The 2 seconds is only a best guess I think - is it enough / any other even in the system we could rely on?

Reply:
Yeah I don't know if @peak3d will be able to respond what the reason for this "minimum 2s" timer thing was.
I kinda get the idea (I think): In case a HDMI_PLUG_EVENT with HDMI: on happened, https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/android/WinSystemAndroid.cpp#L248 checks if the elapsed time is bigger than the "real" set delay of refreshrate and cancles the timer.
If no HDMI_AUDIO_PLUG event appears the timer will fire OnResetDisplay at least after 2s no matter what.

While it probably works for semi-broken boxes where the HDMI_AUDIO_PLUG event is not sent by the OS, it certainly breaks systems where the event works, but the connected AV equipment might take longer than 2s (which might also be rare).

Not sure if there is a need to reflect this 2s minimum in the GUI given above logic.

I tried to keep as much of the logic intact, if my change breaks tons of chinese boxes users have the chance to set delay after refershrate to 100ms which will more or less restore the original behavior.

Reply:
The reason was the following: We might wait for the PLUG Event forever, it might just never come and that this won't happen there is a timeout additionally.

It was once meant to work like on e.g. X11 where one gets an even on unplug and another one on replug - but as said if the second won't come, we would hang there.

Reply:
Additionally, the code logic wanted to be like: If user does not have the feature enabled, trust on the normal even only, but timeout for 2 seconds if event does not come else use whatever user told to do.

So - I think we won't find a timeout that works for everyone. As the default should work though (which is no delay set), just increase the timeout to 5 seconds if delay was not userset? With that chances are better that "out of the box" experience works.

Reply:
Maybe - doing it that way:
- If user has set a timeout -> use this timeout, if hw plug comes first, wait still
- If user has set no timeout, wait for the plug event, but use a large enough timeout, so that we have a chance it works out of the box without annoying the user as user has no idea that we "workaround" his broken setup behind this back?

Keep in mind: There is no perfect solution :-) - so whatever you think is best.

Reply:
Thx for your clarification and the ideas how to do it.
I will think a bit about it, maybe even leave it as is apparently nobody besides me has seen this issue so far...
I assume if it would be common, we would at least see a bunch of github issues about it.

Reply:
Just do it like you think is right :-)

Back at the time, peak had a FireTV Cube or old FireTV or something which was semi-broken and he did not want it to hang :P. 

You care, you decide!

Reply:
The thing is, you can call my test equipment semi-broken as it manages to have HDMI in an off state 2s after the modeset has been done.
So it's hard to argue not sending the event is considered to be "more" broken :-D
I'm really undecided at this point, needs more time and thought definitely.

Reply:
How does your system behave exactly? 

You get the event properly, but **still** need to wait an additional 2 seconds after this event? And the current implementation just "ignores" your setting as the ON comes earlier?

Then I would change this implementation as user setting should always win, beside maybe one difference. We should wait for a certain period of time, when user has it off, as on "proper device" nothing will work until this event comes.

For your device it should behave like this:
User said: 5 seconds
Event comes after 1 - still wait 4.
Event does not come -> return after 5?

Reply:
50% of the time my system works fine just like any other system with following sequence:
1. Playback start
2. Mode get's set
3. onDisplayChange arrives
4. HDMI_AUDIO_PLUG off occurs
5. HDMI_AUDIO_PLUG on occurs

All happens within those 2s and we're fine.

However the other times it goes like:
1. Playback start
2. Mode gets set
3. onDisplayChange arrives (this might happen multiple times for no apparent reason)
4. The 2s timer expires
5. OnResetDislpay gets called
6.  HDMI_AUDIO_PLUG off occurs before AE is initialized
7. AE initializes sink while HDMI is off
8. HDMI_AUDIO_PLUG on occurs at some point
9. Playback starts, I get Timeout in AddData of Audio MediaCodec
10. Playback gets stuck
11. No future playback is possible, because I always end up getting the MediaCodec audio timeout even if I kill the app -> I have to reboot to get it fixed

Reply:
Additional notice:
I also had occasions where I got HDMI OFF, HDMI ON multiple times, but guess there's really the AVR/TV combination to blame.

Reply:
I think I have the same issue on my FireTV 4K when I go through my test video files, where each one changes refreshrate / audio codec. As said before I think here the "user given setting" should in any case be taken into account.

So for your setup, two implementations are maybe a solution
a) Always wait for HDMI ON, but timeout at 5 seconds.
Pro:
- Setups that take sometimes a bit longer now always work. But sometimes you wait 3, and otherwise 2 or max 5 seconds
Cons:
- Semi-broken setups would always wait 5 seconds
b) Always wait at  least the time the user sets in the setting
Pro:
- You set it to 5 seconds and it will always behave the same
Cons:
- It will always need 5 seconds

For jump off / jump on / AE needed to react maybe faster, but as you said -> box needs reboot :-(

From a gut feeling, while I think option a) would be better for your setup. I think option b) is better for a user, cause they can decide by themselves ...

Reply:
```
    if (m_dispResetTimer->IsRunning())
    {
      // We stop the timer if OS supports HDMI_AUDIO_PLUG intent
      // and configured delay is smaller than the time HDMI_PLUG took.
      // Note that timer is always started with minimum of 2 seconds
      // regardless if the configured delay is smaller
      if (m_dispResetTimer->GetElapsedMilliseconds() >=
          CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(
              "videoscreen.delayrefreshchange") *
              100)
        m_dispResetTimer->Stop();
      else
        return; // <- This will cause the RenderIMPl to do this job
    }

    for (auto resource : m_resources)
      resource->OnResetDisplay(); // <- this is only triggered from here when the display ON comes back and timer is gone. Else the timer would timeout at a point and it is triggered from: WinSystemAndroidGLESContext.cpp
    m_dispResetTimer->Stop();
    m_HdmiModeTriggered = false;
  }
```

I am wondering if this issue is actually force triggered xbmc/windowing/android/WinSystemAndroidGLESContext.cpp which causes nothing to wait, whenever no timer is set?

It seems the timer is once set from XBMCApp.cpp whenever the system gets the PLUG event, then it is checked if the "user set" timer is already gone, if not - it just returns. Then the RenderIMpl from above always checks if the timer is still runing (per frame) and tries to wakeup the resources.

So, the 2 seconds is actually a workaround for RenderIMpl - but I think your setup should already work right now - if you set it to 5 seconds? Right?

Reply:
Yeah I assume 5s will be fine for my setup, unless I stumble across runs where it takes even longer.
Will do some more testing and collect some logs of fail runs.